### DISCUSSION TITLE
merge to shared-dev

### DISCUSSION THREAD

**Discussion by markusschloesser** [markusschloesser/MackieC4_P3#80] [General]
I merged master to shared-dev, so now master, shared-dev and encodermaphistory branches should all be identical.
I will continue to develop in shared-dev for the minor things, while the componentguard branch is still a major mess and experimentation playground for the NEW stuff :-)

**Comment by BilldarBagdar** [comment]
ok cool.  I have not pulled down any updates to my local repo in a week now maybe.  I will let you know if I notice anything "disappear" again when I fetch-and-merge.  Hopefully, you've figured out why it happened before and preempted it from happening again. :)  

FYI - I suspect the "problem" is/was that git always generally expects a developer to be "up to date with origin" or something.  The scenario I'm thinking of goes like this...  If we are each working in our own branches and happen to change the same file and, I commit a change to that file in my branch on "day 01" while you commit a change in your branch on "day 02", then I push my "day 01" commit to origin on "day 03"... Then, I think, whenever you Pull my "day 01" commit from origin there is a chance that some "day 01" code from "my commit" gets spuriously masked by your "day 02" commit of a change to the same file.

Also FYI - The way I have found these "missing commit fragments" is by doing "fetch from origin" first (before Pulling) and comparing all the differences with my local branch copies.  Now that I am more familiar with the GUI in SourceTree, it's a pretty easy comparison to perform.  After fetching from origin, right click on the "other branch" (under Remotes/origin) and select "Diff against Current...", and then any "missing fragments" will be "red" in the diff display boxes (meaning these lines in the checked out branch are "missing from the 'other' branch" in my case and the opposite, these lines in the 'other' branch are "missing from the 'checked out' branch" in your case). If/when those "red lines" disappear inappropriately after a Pull, they are only recoverable if you know they exist and recover them.

**Reply by markusschloesser** [reply]
> ok cool. I have not pulled down any updates to my local repo in a week now maybe. I will let you know if I notice anything "disappear" again when I fetch-and-merge. Hopefully, you've figured out why it happened before and preempted it from happening again. :)

I hope so, too! But I think so.

> Also FYI - The way I have found these "missing commit fragments" is by doing "fetch from origin" first (before Pulling) and comparing all the differences with my local branch copies. Now that I am more familiar with the GUI in SourceTree, it's a pretty easy comparison to perform. After fetching from origin, right click on the "other branch" (under Remotes/origin) and select "Diff against Current...", and then any "missing fragments" will be "red" in the diff display boxes (meaning these lines in the checked out branch are "missing from the 'other' branch" in my case and the opposite, these lines in the 'other' branch are "missing from the 'checked out' branch" in your case). If/when those "red lines" disappear inappropriately after a Pull, they are only recoverable if you know they exist and recover them.

That's exactly what I do! :-)

**Reply by BilldarBagdar** [reply]
I didn't do an exhaustive comparison this time because everything was already merged into shared-dev, I just didn't update the other branches and pulled shared-dev straight in after a quick glance at the after-fetch diff.  Then I "pushed it" over to Live and started test running.  Depending on the order in which you read my recent (string of back-to-back?) posts, you may or may not already know more-or-less the scope of the test running I've done.  No disappeared code encountered. (and none expected, but there's always that sliver of "proving some unknown does NOT exist" doubt either symbiotically stabilizing or parasitically leeching off of the vine of certainty depending on one's perspective. LOL How poetic? I mean, at this point I'm going to gamble on it. Rather than actually performing "all" the close comparison work to prove nothing disappeared.  I'll defer paying "all" that cost, if any, until some evidence appears pointing that way.)  :)

**Comment by BilldarBagdar** [comment]
Hi Markus,

I'm trying to Fetch and Pull from this repository, but github is telling me "authentication rejected", and "Please make sure you have the correct access rights and the repository exists."

Do I still have access rights in this repository?  If so, I need to keep looking for the issue on my local end.

Cheers.  I noticed you've made some updates over the summer.  I finally updated to Live 11, although I'm still running two copies at version 10 because those are Windows 7 machines.

**Reply by markusschloesser** [reply]
Hi! Let me check, I think I was trying to improve the security and limited something.
**right now reading resources on scripting :-) **

**Reply by markusschloesser** [reply]
you are still listed as collaborator (of course!), the only thing I changed was enabling 
"Require a pull request before merging.
When enabled, all commits must be made to a non-protected branch and submitted via a pull request before they can be merged into a branch that matches this rule."
and
"Require approvals
When enabled, pull requests targeting a matching branch require a number of approvals and no changes requested before they can be merged."
But I didn't thing that would affect collaborators as well??
I have now disabled the latter, please try again.
And hey! :-) Welcome back!

**Comment by BilldarBagdar** [comment]
Thanks, that wasn't the issue.  You are probably safe to re-enable those "Pull Request" security settings.  

Are there any other collaborators yet?  At work, we generally used the "Pull Request before Merge" and "required approvals" protection as a "stopping point" in the development cycle for "code review".  If only owners and collaborators can Merge anyway, and I'm still the only collaborator, maybe such branch integrity protection might be overkill?

**Reply by BilldarBagdar** [reply]
I successfully pulled 8 commits down to my local master.  I had to register "this local machine's" SSH public key in Bildar Bagdar's GitHub account again.  I guess I deleted it last April when I set up an SSH key for a different local machine.

**Reply by markusschloesser** [reply]
👍☺ 
you should start with 2022_dev.
That includes quite a bit of new stuff, docstrings and fixes

**Reply by markusschloesser** [reply]
unfortunately no other collaborators. 
I also didn't change much in the script for a year or so because it was just working and a joy to work with. Only when I had some nasty bugs, i finally started again. (that was the metering in `on_update_display_timer `which wasn't needed anyway)
The last couple of weeks I really dug in again and started 
cleaning up, 
disabling some unnecessary logging, 
and implemented some new things (mainly Solo Button in ChannelStrip mode, Beat/Song Position pointer in Function mode, which unfortunately currently only SHOWS the time, but doesn't change it yet (I want to have a Jog Dial on the Encoder, with the Vpot ring showing "how far" in the track I am).
Also fixed a couple of not properly working vpot unlight/enlight by moving it to `on_update_display_timer`

**Reply by BilldarBagdar** [reply]
thanks, I'll take a look at what's new.  

Did shared-dev go stale or die?  It looks like my two most recent commits (1.5 ish years ago? really?) were for troubleshooting and remain at the tip of that branch.

**Reply by BilldarBagdar** [reply]
Ha! Yes, the script does work well already.  I have not exercised it much recently, not "hard", but likewise haven't noticed any problems using it since we got things basically stabilized.

**Reply by markusschloesser** [reply]
> thanks, I'll take a look at what's new.
> 
> Did shared-dev go stale or die? It looks like my two most recent commits (1.5 ish years ago? really?) were for troubleshooting and remain at the tip of that branch.

nope, I just didn't want to mess up our shared branch, so I started a new one.
But I am happy to push an update tomorrow to 2022_dev and after that, we could "merge" to master. Then I am free to mess things up again.
Oh and also, we merged shared-dev to master iirc, so it shouldn't be ahead?

**Comment by BilldarBagdar** [comment]
I think I've pulled all the latest updates down, but I will double check.

Just want to add, don't be afraid to prune branches either.  There's nothing sacred about branches in git because of the commit history.  I no longer immediately remember why the "encoder map history" branch was created, for example, but if the branch contents are now exactly the same as master and shared-dev, can it be deleted, pruned from the "living," active tree?
